### ISSUE TITLE
when switching to required suffix and/or prefix, can't save customers any longer

### ISSUE THREAD

**Issue by ludwig-gramberg** [magento/magento2#7499]
when configuring magento to require either suffix or prefix it's no longer possible to save customers which do not have the (now missing) values yet.
the customer model and its related addresses are being validated in the old state when the customer is saved before the new address data is supposed to be saved, thus making it impossible to ever meet the validation requirement for the addresses.

### Preconditions
1. magento 2.1.2

### Steps to reproduce
1. set magento to not require a suffix
2. in adminpanel save a customer and a customer address without suffix
3. configure magento to require a suffix
4. try to save the customer now providing a suffix for both the customer as well as his address

### Expected result
1. the customer saves and now has a suffix 
2. the customer saves and his address has a suffix

### Actual result
1. an error message appears: "suffix is a required field"

### Details

Validation occurs in:
`Magento\Customer\Model\ResourceModel\Address` at line 89
originating from
`Magento\Customer\Model\ResourceModel\Customer\Relation` in method `processRelation` at line 49
the address object being passed to this stems from
`Magento\Customer\Model\ResourceModel\CustomerRepository` in method `save` at line 149 to 158: 
```
[...]

$origAddresses = $customer->getAddresses();
$customer->setAddresses([]);
$customerData = $this->extensibleDataObjectConverter->toNestedArray(
      $customer,
      [],
      '\Magento\Customer\Api\Data\CustomerInterface'
);

$customer->setAddresses($origAddresses);
$customerModel = $this->customerFactory->create(['data' => $customerData]);

[...]
```

as you can see the customer factory is creating the customerModel from customerData which was extracted from a customer object with no addresses, so later when that model is begin saved, its validated against its old version of addresses which do not have a suffix value yet.
if the addresses are being validated in relation to the save of the customer, then the addresses should be saved first thus having a chance to be valid... right? or just don't automatically validate them at all when the customer is being saved...

**Comment by sevos1984** [comment]
Can't reproduce this. After step 4 customer is saved and has both suffixes saved. Maybe you have some other options enabled? Have you tried clearing magento and browser cache?

**Comment by ludwig-gramberg** [comment]
we upgraded from magento 2.0.7 to 2.1.2 maybe it has to do with that? does your customer has addresses? the exception comes from the addresses being validated when the customer is being saved and not having the prefix/suffix yet (will happen only _after_ the customer would be saved but based on his old addresses)

**Comment by sevos1984** [comment]
Reproduced that in 2.1.2 but not in dev, seems like fixed already.

**Comment by sevos1984** [comment]
Fix should be included in next patch, reopen if that doesn't help

**Comment by ludwig-gramberg** [comment]
will do, thx!

**Comment by chrisgan** [comment]
I am seeing this same issue on version 2.1.6

With a fresh database, I created one customer before enabling the 'Prefix' option as a required field.
Afterwards, when checking out, I am redirected to the Edit Account Information screen with an error stating 'prefix is a required field'. When I then select a prefix from this page and save, it does not save and returns an error again.

**Comment by ludwig-gramberg** [comment]
this appears to be fixed in magento 2.1.7
wasnt able to reproduce this any more
@chrisgan can you confirm?

@sevos1984 can you point me to the commit so i can make sure?

**Comment by chrisgan** [comment]
@ludwig-gramberg 
@sevos1984 

I have just tested this on 2.1.7, fresh install on a local environment. I've also noticed another issue with this since with password reset. 

These are the tests I did:

Create an account without a prefix, as the prefix field is not currently enabled.
Set Prefix as a Required Field in the admin

As a customer without a prefix, do the following:

Try to reset password via password reset form
Receive error 'We're unable to send the password reset email.'
Error in system.log is 'Exception message: "Title" is a required value.'

Or, assuming you are logged in as the customer without a prefix:

Try to checkout
Get redirected to the account page
Receive error 'The value of attribute "prefix" must be set
(note that is also a bug in the translation, as we have Prefix translated as 'Title'
Try to save the account with a prefix
Receive error "Title" is a required value. (translation is working here)

I don't believe the issue is resolved based on my testing.

**Comment by ludwig-gramberg** [comment]
@chrisgan i tested saving the customer in admin panel and changing the address by the customer in his account, there it worked, will test checkout and password reset too

**Comment by chrisgan** [comment]
@ludwig-gramberg further to this, we also saw the error on some accounts when they attempted to login, it may be because Magento was saving information against the account at this point as well, but did not manage to recreate this locally.

Essentially, it seems the trouble is, once the field is required, the account model can not be saved anymore, without a Prefix being set. But of course this is particularly difficult to resolve if the customer is not logged in, and having the addresses there as well makes this worse, as they all need to have the Prefix set as well.

**Comment by ludwig-gramberg** [comment]
@chrisgan my orignal issue was though that, you couldn't save an address at all any more after changing the setting, because it would validate against the old address first which never met the new criteria

**Comment by chrisgan** [comment]
@ludwig-gramberg I think the original issue as described though 'when switching to required suffix and/or prefix, can't save customers any longer' is still a problem, whilst it is possible to save the addresses, enabling the prefix does still break the existing customer accounts.

Do you experience the issues I described above?

@sevos1984 would you suggest I raise this as a separate issue? Or would you suggest re-opening this ticket?

**Comment by ludwig-gramberg** [comment]
@chrisgan turns out i didn't have suffix/prefix configured as required fields as I did originally, now that i changed it the behaviour is as before

**Comment by ludwig-gramberg** [comment]
@chrisgan btw if you set it to optional but you define a list of values you basically have the desired behaviour, the user can not leave the field empty since its a select. only thing to consider is that a person could craft a request where the data is missing...

**Comment by chrisgan** [comment]
@ludwig-gramberg thanks for the feed back on this.

Actually I am not including the semi-colon at the start of the options, but the drop down still starts empty.

There is also the problem with customers with existing accounts, I think these customers should be required to update their details, but only in a friendly way during checkout.

And of course, having it not shown as required is not a good experience from a customer experience point of view.

At the moment, the only real solution is to update all previous accounts so they have a value stored for the Prefix, I would suggest the issue is re-opened @sevos1984
